World Bank (Independent Publisher)

krautrocker opened the pull request:
---
### When submitting a connector, please make sure that you follow the requirements below, otherwise your PR might be rejected. We want to make you have a well-built connector, a smooth certification experience, and your users are happy :) 

If this is your first time submitting to GitHub and you need some help, please sign up for this [session](https://forms.office.com/pages/responsepage.aspx?id=KtIy2vgLW0SOgZbwvQuRaXDXyCl9DkBHq4A2OG7uLpdUMTFJWFFGVUxBNUFZQjZWRUdaOE5BMFkwNS4u). 

- [x] I attest that the connector doesn't exist on the Power Platform today. I've verified by checking the pull requests in GitHub and by searching for the connector on the platform or in the documentation. 
- [x] I attest that the connector works and I verified by deploying and testing all the operations. 
- [x] I attest that I have added detailed descriptions for all operations and parameters in the swagger file.
- [x] I attest that I have added response schemas to my actions, unless the response schema is dynamic. 
- [x] I validated the swagger file, `apiDefinition.swagger.json`, by running `paconn validate` command.
- [x] If this is a certified connector, I confirm that `apiProperties.json` has a valid brand color and doesn't use an invalid brand color, `#007ee5` or `#ffffff`. If this is an independent publisher connector, I confirm that I am not submitting a connector icon.

If you are an Independent Publisher, you must also attest to the following to ensure a smooth publishing process:
- [x] I have named this PR after the pattern of "Connector Name (Independent Publisher)" ex: HubSpot Marketing (Independent Publisher)
- [x] Within this PR markdown file, I have pasted screenshots that show: 3 unique operations (actions/triggers) working within a Flow. This can be in one flow or part of multiple flows. For each one of those flows, I have pasted in screenshots of the Flow succeeding. 
- [x] Within this PR markdown file, I have pasted in a screenshot from the Test operations section within the Custom Connector UI.
- [x] If the connector uses OAuth, I have provided detailed steps on how to create an app in the readme.md.

krautrocker committed:
- Initial commit.

krautrocker commented:
Testing:

![Screenshot 2025-06-02 222007](https://github.com/user-attachments/assets/6b314351-84f0-4cd6-a9ec-3c954ff8e4e0)

![image](https://github.com/user-attachments/assets/dd787a44-7989-4151-9fc4-95f3740f83db)

krautrocker commented:
Flow Proof-of-Life:

**Debt-to-GDP:**

![Screenshot 2025-06-02 222218](https://github.com/user-attachments/assets/552024af-30a6-42dc-a7c9-d1660ed41c43)

**Globa Exports:**

![Screenshot 2025-06-02 222200](https://github.com/user-attachments/assets/3bc55bbf-7504-4138-a8f2-2a344981282b)

**Data Sources Reference List:**

![Screenshot 2025-06-02 222143](https://github.com/user-attachments/assets/a7e67e8f-3307-482a-9dc3-6efcecb7f573)

krautrocker committed:
- Removed per_page parameter

krautrocker commented:
Adding parameters tomorrow:

- page
- per_page
- format
- date
- mrv

krautrocker committed:
- Added additional query string parameters to each endpoint.

- paconn validated, no errors.

krautrocker committed:
- Added error code 503 to responses

krautrocker commented:
This is one way to process the results:

- Full body: Compose -> body('Get_All_Sources')
- Metadata (pages, counts): body('Get_All_Data_Sources')**[0]**
- Data array: body('Get_All_Data_Sources')**[1]**

![image](https://github.com/user-attachments/assets/06c91d7e-43ea-4146-ac6f-e89a47716d66)

krautrocker commented:
WB endpoints require pagination - it's complicated because two arrays are returned, one for metadata and another for the actual data. More to come on this by end of the week.

certificationteam commented:
> WB endpoints require pagination - it's complicated because two arrays are returned, one for metadata and another for the actual data. More to come on this by end of the week.

Hi @krautrocker, 

Thank you for the update, please give us a heads-up when this PR is completed.

krautrocker committed:
- Updated readme and definition to include guidance on paginating the response.

krautrocker commented:
@certificationteam Just made some updates to the definition and readme files. They point to documentation that I put together detailing how to paginate the responses.

The endpoints don't offer pagination natively and the output structure requires a truly custom approach. This works with all endpoints offered in the connector.

certificationteam commented:
> @certificationteam Just made some updates to the definition and readme files. They point to documentation that I put together detailing how to paginate the responses.
> 
> The endpoints don't offer pagination natively and the output structure requires a truly custom approach. This works with all endpoints offered in the connector.

Thank you @krautrocker for update.  Now is this PR ready for review?

troystaylor commented:
Sorry @krautrocker, I meant to reply earlier. I have custom code you can use to handle the response array with mixed types: https://github.com/troystaylor/Connector-Code/blob/main/ArrayResponseMixedTypes.csx
You will need to change the format parameter to default to JSON (and probably x-ms-visibility: internal), then define your 200 responses with the new schema.
![image](https://github.com/user-attachments/assets/5155054f-e821-4906-a2ee-d92cd19152f8)

krautrocker commented:
@certificationteam Yes, this is ready. I'm going to have a look at what @troystaylor provided and try to integrate this into the connector.

krautrocker commented:
@troystaylor All good - thanks for providing and giving it a test. I'll review this and integrate it into the connector (with credit, of course). There's one thing that I need to add but hoping this makes it more straightforward, it was a surly beast.

krautrocker commented:
I ran a number of tests - the C# works nicely, thanks a lot. It doesn't handle all the pages straight away, so I added pagination on my end. I've also added query string support for format, date, page, per_page and mrv.

Pausing to integrate this right away because of the path parameters. This complicates things; the script grows considerably to accommodate them, and the only way around them makes me think it may not be scalable.

I've got something cooking in C# - I'll post over the weekend after I finish testing it out. Thanks again @troystaylor

krautrocker commented:
Proof-of-life for paging script that I'll post later - if you look at the second page of results for GetAllSources, these are the results from the second page. Results from all pages are now included in a single response.

![image](https://github.com/user-attachments/assets/0fda9259-d34e-455e-b27b-b3d4513a70cc)

krautrocker commented:
@certificationteam I've released an article on this subject with credit to @troystaylor on the C# portion. There are a few ways to handle the responses now, as highlighted [here](https://developments.substack.com/p/connecting-the-goats).

I've tested this with every endpoint and the Mondo Aggregator option (#2) works for all of them, even when no criteria are present to filter the response. So, I'm tempted to submit Mondo Aggregator as the custom code for review and release to the public, assuming everything is copacetic. Page Slayer processes each page but a Do Until loop is still required; implementation details are available in the article in the event that the Mondo Aggregator approach proves troublesome or problematic.

Currently, the connector code does not implement either of these methods, but I will submit changes to the definition after deliberating which method is the best option for users at large. Open to guidance on this, of course.
